llvm/include/llvm/AsmParser/AsmParserState.h

```diff
@@ -3,9 +3,26 @@
 
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/IR/Value.h"
+#include <optional>
 
 namespace llvm {
-struct AsmParserState {
+class AsmParserState {
+public:
+  std::optional<FileLocRange> getFunctionLocation(const Function *) const;
+  std::optional<FileLocRange> getBlockLocation(const BasicBlock *) const;
+  std::optional<FileLocRange> getInstructionLocation(const Instruction *) const;
+  std::optional<Function *> getFunctionAtLocation(const FileLocRange &) const;
+  std::optional<Function *> getFunctionAtLocation(const FileLoc &) const;
+  std::optional<BasicBlock *> getBlockAtLocation(const FileLocRange &) const;
+  std::optional<BasicBlock *> getBlockAtLocation(const FileLoc &) const;
+  std::optional<Instruction *>
+  getInstructionAtLocation(const FileLocRange &) const;
+  std::optional<Instruction *> getInstructionAtLocation(const FileLoc &) const;
+  bool addFunctionLocation(Function *, const FileLocRange &);
+  bool addBlockLocation(BasicBlock *, const FileLocRange &);
+  bool addInstructionLocation(Instruction *, const FileLocRange &);
+
+private:
   DenseMap<Function *, FileLocRange> Functions;
   DenseMap<BasicBlock *, FileLocRange> Blocks;
   DenseMap<Instruction *, FileLocRange> Instructions;
```

Comment:
I am still of the opinion that we need to use `SmallVector<std::pair<Function *, FileLocRange>` here(and similarly for Blocks, and Instructions). My main motivation for this change is that, most of our queries are going to ask for a Function/BasicBlock/Instruction given a Location in the file, which is going to force us to iterate through DenseMap and find the IR component asked. If anyway we are forced to iterate, a SmallVector is a better option, as it would atleast give better cache locality. 

If you still think DenseMap would be better here, it would be helpful if you state out your reasoning.

Reply:
On the other hand, I think I might be asking you to do this optimization prematurely. So feel free to ignore.